include/unifex/when_all.hpp

```diff
@@ -213,19 +213,39 @@ struct _op<Receiver, Senders...>::type {
 
   void start() noexcept {
     stopCallback_.construct(
-        get_stop_token(receiver_), cancel_operation{stopSource_});
+        get_stop_token(receiver_),
+        cancel_operation<Receiver, Senders...>{*this});
     ops_.start();
   }
 
-  private:
+  void request_stop() noexcept {
+    // mark callback as running (own deliver_result)
+    auto oldCount = refCount_.fetch_add(1, std::memory_order_release);
+    UNIFEX_ASSERT(!(oldCount & callback_running_bit));
+    stopSource_.request_stop();
+    // unmark (deliver_result back to operation)
+    oldCount = refCount_.fetch_sub(1, std::memory_order_acq_rel);
+    UNIFEX_ASSERT(oldCount & callback_running_bit);
+    // deliver_result as there are no more operations
+    if (oldCount == 1) {
+      unifex::set_done(std::move(receiver_));
+    }
+  }
+
+private:
   void element_complete() noexcept {
-    if (refCount_.fetch_sub(1, std::memory_order_acq_rel) == 1) {
+    // +2 for last operation, +1 for callback running
+    if (refCount_.fetch_sub(2, std::memory_order_acq_rel) < 4) {
```

Comment:
You should be able to relax this to just a release operation if you make the below load in `deliver_result` an acquire operation. 

As is, I'm not sure how other threads would be able to take advantage of this acquire operation (no way to determine which threads have already completed).

Reply:
ASAN problem was only triggered in a single threaded scenario when `request_stop` resulted in calling `set_done()` inline. If another thread 'wins' the `deliver_result`, it would be blocked on `stopCallback_.destruct()`.